_posts/2024-03-03-pubsub-consumers.md

```diff
@@ -0,0 +1,57 @@
+
+
+![](/assets/amqp-clients.jpg)
+In this diagram, each logical application has its own message queue. Using the "competing consumers" model, exactly one client instance is able to consumer a message per logical application.
+
+
+One downside of this approach is that it makes broadcasts more difficult. If there are some messages that should be received by all clients, each client requires an additional, dedicated queue (this is the reason NServiceBus doesn't support broadcast events and instead recommends other approaches).
+
+## Publish/Subscribe with logs
+
+Event Stream Processing is another approach to messaging that has gained a lot of popularity. Kafka is probably the best example, but both AWS and Azure offer similar products as a service. I'm using Kafka as the technology name here, as it's so commonly used.
+
+Kafka publishers store messages in a topic partition. Every topic has one or more partitions which persistes the messages. Subscribers can then read messages from a partition at any point without the message being removed. Kafka also defines the concept of client groups, which ultimately is the same as our "logical client". Each client group can have multiple client instances that belong to the group.
+
+In kafka, a partition can only be read by one client instance within the same client group. This has several implications:
+* Similar to "traditional" messages queues, one event will be processed by a single client instance within the logical application.
+* It's harder to mix broadcast and competing consumers models at the same time because a client instance belongs to a single client group and broadcasts effectly only work across client groups, not across client-instances.
+* The maximum number of client instances is limited by the number of partitions of a topic as only one client intance can process a partition. This limits scalability (also scaling up or down is more complex as client instances need to be rebalanced across the partitions).
+* Load balancing primarily depends on the partitioning strategy and the number of partitions. Over-simplified: Scaling subscriber instances is more difficult than using traditional message queuing technologies.
+
+![](/assets/kafka-clients.jpg)
+The diagram shows how each subscriber instance of a logical instance is assigned to one more topic partitions. A publisher publishes an event to a specific partition which will be read by one subscriber instance per logical application.
+
+An additional difference of Kafka is that messages aren't removed from the broker once a client has processed it. Therefore, clients can decide to re-process messages.
```

Comment:
```suggestion
Another difference in Kafka is that messages aren't removed from the broker once a client has processed them. Therefore, clients can decide to re-process messages.
```

Is it worth also mentioning that clients need to maintain a pointer to where they were?

Reply:
> Is it worth also mentioning that clients need to maintain a pointer to where they were?

hm maybe as a sidenote. I think it doesn't really have a meaningful impact on the topic being brough up here?